File: backend/src/services/PasswordService.ts
import bcrypt from 'bcrypt';
import crypto from 'crypto';
import { env } from '../config/env';

// In-memory session store (replace with Redis in production)
const sessionStore = new Map<string, { uid: string; expiresAt: number }>();

/**
 * Hash a password using bcrypt with configured cost factor.
 */
export async function hashPassword(password: string): Promise<string> {
  return bcrypt.hash(password, env.BCRYPT_ROUNDS);
}

/**
 * Verify a password against a bcrypt hash.
 */
export async function verifyPassword(password: string, hash: string): Promise<boolean> {
  return bcrypt.compare(password, hash);
}

/**
 * Create a session token for authenticated access to a private share.
 * Token expires after 15 minutes.
 */
export function createSessionToken(uid: string): string {
  const token = crypto.randomBytes(32).toString('hex');
  const expiresAt = Date.now() + 15 * 60 * 1000; // 15 minutes

  sessionStore.set(token, { uid, expiresAt });
  return token;
}

/**
 * Validate a session token. Returns the associated UID if valid.
 */
export function validateSessionToken(token: string, uid: string): boolean {
  const session = sessionStore.get(token);
  if (!session) return false;

  // Check expiry
  if (Date.now() > session.expiresAt) {
    sessionStore.delete(token);
    return false;
  }

  // Check that token is for the correct UID
  if (session.uid !== uid) return false;

  return true;
}

/**
 * Cleanup expired sessions (called periodically).
 */
export function cleanupExpiredSessions(): void {
  const now = Date.now();
  for (const [token, session] of sessionStore.entries()) {
    if (now > session.expiresAt) {
      sessionStore.delete(token);
    }
  }
}

// Run session cleanup every 5 minutes
setInterval(cleanupExpiredSessions, 5 * 60 * 1000);

File: backend/app/services/password_service.py
"""
password_service.py — bcrypt hashing + in-memory session tokens.
Replaces: backend/src/services/PasswordService.ts
"""

import secrets
import time
import threading

from passlib.hash import bcrypt

from app.config import settings

# In-memory session store — same structure as the Node.js Map<string, {uid, expiresAt}>
_session_store: dict[str, dict] = {}
_lock = threading.Lock()


async def hash_password(password: str) -> str:
    """Hash a password using bcrypt with configured rounds."""
    return bcrypt.using(rounds=settings.BCRYPT_ROUNDS).hash(password)


async def verify_password(password: str, hashed: str) -> bool:
    """Verify a password against a bcrypt hash."""
    return bcrypt.verify(password, hashed)


def create_session_token(uid: str) -> str:
    """Create a session token for authenticated access to a private share (15-min TTL)."""
    token = secrets.token_hex(32)
    expires_at = time.time() + 15 * 60  # 15 minutes
    with _lock:
        _session_store[token] = {"uid": uid, "expires_at": expires_at}
    return token


def validate_session_token(token: str, uid: str) -> bool:
    """Validate a session token. Returns True if valid for the given UID."""
    with _lock:
        session = _session_store.get(token)
        if session is None:
            return False
        if time.time() > session["expires_at"]:
            del _session_store[token]
            return False
        if session["uid"] != uid:
            return False
        return True


def cleanup_expired_sessions() -> None:
    """Remove expired sessions from the in-memory store."""
    now = time.time()
    with _lock:
        expired = [k for k, v in _session_store.items() if now > v["expires_at"]]
        for k in expired:
            del _session_store[k]
